[receiver/sqlquery] Broken Test TestPostgresIntegrationLogsTrackingWithStorage

### Component(s)

receiver/sqlquery

### What happened?

Since https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31667 was merged the `TestPostgresIntegrationLogsTrackingWithStorage` integration test has started failing silently for some reason.

You can see the failures inside the `Run Integration Test` step in jobs like:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8254951071/job/22580335315
- https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8268349528/job/22621241890
any another other job since.

![image](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/12352919/9dc33f4e-68e2-48dc-b7ef-946696667954)

For some reason my `update-otel` PR is actually showing the failure: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8300689476/job/22720973013

### Collector version

Latest

### Environment information

_No response_

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

Thanks for filing, I'll take a look when I get a chance 👍 

Pasting the test failure output:
```
=== FAIL: . TestPostgresIntegrationLogsTrackingWithStorage (10.80s)
2024/03/15 19:21:26 🐳 Creating container for image postgres:9.6.24
2024/03/15 19:21:26 ✅ Container created: 8b4be6dec43e
2024/03/15 19:21:26 🐳 Starting container: 8b4be6dec43e
2024/03/15 19:21:26 ✅ Container started: 8b4be6dec43e
2024/03/15 19:21:26 🚧 Waiting for container id 8b4be6dec43e image: postgres:9.6.24. Waiting for: &{Port:5432 timeout:0xc000bee0f8 PollInterval:100ms}
{"level":"debug","msg":"starting..."}
{"level":"debug","msg":"started."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopping..."}
{"level":"debug","msg":"stopped."}
{"level":"debug","msg":"starting..."}
make[2]: *** [../../Makefile.Common:142: mod-integration-test] Error 1
make[1]: *** [Makefile:165: receiver/sqlqueryreceiver] Error 2
make: *** [Makefile:122: gointegration-test] Error 2
{"level":"debug","msg":"started."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopping..."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopped."}
    integration_test.go:205: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/sqlqueryreceiver/integration_test.go:205
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 5
        	Test:       	TestPostgresIntegrationLogsTrackingWithStorage
2024/03/15 19:21:37 🐳 Terminating container: 8b4be6dec43e
2024/03/15 19:21:37 🚫 Container terminated: 8b4be6dec43e

=== FAIL: . TestPostgresIntegrationLogsTrackingWithStorage (re-run 1) (11.37s)
2024/03/15 19:22:08 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 24.0.9
  API Version: 1.43
  Operating System: Ubuntu 22.04.4 LTS
  Total Memory: 15981 MB
  Resolved Docker Host: unix:///var/run/docker.sock
  Resolved Docker Socket Path: /var/run/docker.sock
  Test SessionID: [255](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8300689476/job/22720973013#step:5:256)3bd2f08ff28a5fd5182f13a3edcdd5e001a7d0df1549d51ff6a924ce069e3
  Test ProcessID: bc91307c-2e72-4b5c-8d69-17e4ca4f2c6f
2024/03/15 19:22:08 🐳 Creating container for image testcontainers/ryuk:0.6.0
2024/03/15 19:22:08 ✅ Container created: 162027f87ee3
2024/03/15 19:22:08 🐳 Starting container: 162027f87ee3
2024/03/15 19:22:08 ✅ Container started: 162027f87ee3
2024/03/15 19:22:08 🚧 Waiting for container id 162027f87ee3 image: testcontainers/ryuk:0.6.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms}
2024/03/15 19:22:08 🐳 Creating container for image postgres:9.6.24
2024/03/15 19:22:08 ✅ Container created: b71564ce7d82
2024/03/15 19:22:08 🐳 Starting container: b71564ce7d82
2024/03/15 19:22:08 ✅ Container started: b71564ce7d82
2024/03/15 19:22:08 🚧 Waiting for container id b71564ce7d82 image: postgres:9.6.24. Waiting for: &{Port:5432 timeout:0xc00078c5c8 PollInterval:100ms}
{"level":"debug","msg":"starting..."}
{"level":"debug","msg":"started."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopping..."}
{"level":"debug","msg":"stopped."}
{"level":"debug","msg":"starting..."}
{"level":"debug","msg":"started."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopping..."}
{"level":"debug","msg":"Running query"}
{"level":"debug","msg":"stopped."}
    integration_test.go:205: 
        	Error Trace:	/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/receiver/sqlqueryreceiver/integration_test.go:205
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 5
        	Test:       	TestPostgresIntegrationLogsTrackingWithStorage
2024/03/15 19:22:19 🐳 Terminating container: b71564ce7d82
2024/03/15 19:22:19 🚫 Container terminated: b71564ce7d82

DONE 2 runs, 6 tests, 1 skipped, 2 failures in 105.373s
make[1]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib'
Error: Process completed with exit code 2.
```

--- a/.chloggen/sqlquery_shutdown.yaml
+++ b/.chloggen/sqlquery_shutdown.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: sqlqueryreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix memory leak on shutdown for log telemetry
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31782]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/receiver/sqlqueryreceiver/go.mod
+++ b/receiver/sqlqueryreceiver/go.mod
@@ -20,7 +20,6 @@ require (
 	go.opentelemetry.io/otel/metric v1.24.0
 	go.opentelemetry.io/otel/trace v1.24.0
 	go.uber.org/goleak v1.3.0
-	go.uber.org/multierr v1.11.0
 	go.uber.org/zap v1.27.0
 )
 
@@ -141,6 +140,7 @@ require (
 	go.opentelemetry.io/otel/exporters/prometheus v0.46.0 // indirect
 	go.opentelemetry.io/otel/sdk v1.24.0 // indirect
 	go.opentelemetry.io/otel/sdk/metric v1.24.0 // indirect
+	go.uber.org/multierr v1.11.0 // indirect
 	golang.org/x/crypto v0.21.0 // indirect
 	golang.org/x/exp v0.0.0-20240205201215-2c58cdc269a3 // indirect
 	golang.org/x/mod v0.14.0 // indirect

--- a/receiver/sqlqueryreceiver/integration_test.go
+++ b/receiver/sqlqueryreceiver/integration_test.go
@@ -24,6 +24,7 @@ import (
 	"go.opentelemetry.io/collector/component/componenttest"
 	"go.opentelemetry.io/collector/consumer/consumertest"
 	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver"
 	"go.opentelemetry.io/collector/receiver/receivertest"
 	"go.uber.org/zap"
 
@@ -48,7 +49,8 @@ func TestPostgresIntegrationLogsTrackingWithoutStorage(t *testing.T) {
 	}()
 
 	// Start the SQL Query receiver.
-	receiver, config, consumer := createTestLogsReceiverForPostgres(t, externalPort)
+	receiverCreateSettings := receivertest.NewNopCreateSettings()
+	receiver, config, consumer := createTestLogsReceiverForPostgres(t, externalPort, receiverCreateSettings)
 	config.CollectionInterval = time.Second
 	config.Queries = []sqlquery.Query{
 		{
@@ -84,7 +86,7 @@ func TestPostgresIntegrationLogsTrackingWithoutStorage(t *testing.T) {
 	require.NoError(t, err)
 
 	// Start new SQL Query receiver with the same configuration.
-	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort)
+	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort, receiverCreateSettings)
 	config.CollectionInterval = time.Second
 	config.Queries = []sqlquery.Query{
 		{
@@ -134,7 +136,8 @@ func TestPostgresIntegrationLogsTrackingWithStorage(t *testing.T) {
 	storageExtension := storagetest.NewFileBackedStorageExtension("test", storageDir)
 
 	// create SQL Query receiver configured with the File Storage extension
-	receiver, config, consumer := createTestLogsReceiverForPostgres(t, externalPort)
+	receiverCreateSettings := receivertest.NewNopCreateSettings()
+	receiver, config, consumer := createTestLogsReceiverForPostgres(t, externalPort, receiverCreateSettings)
 	config.CollectionInterval = time.Second
 	config.StorageID = &storageExtension.ID
 	config.Queries = []sqlquery.Query{
@@ -176,7 +179,7 @@ func TestPostgresIntegrationLogsTrackingWithStorage(t *testing.T) {
 	testAllSimpleLogs(t, consumer.AllLogs())
 
 	// start the SQL Query receiver again
-	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort)
+	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort, receiverCreateSettings)
 	config.CollectionInterval = time.Second
 	config.StorageID = &storageExtension.ID
 	config.Queries = []sqlquery.Query{
@@ -209,7 +212,7 @@ func TestPostgresIntegrationLogsTrackingWithStorage(t *testing.T) {
 	insertPostgresSimpleLogs(t, dbContainer, initialLogCount, newLogCount)
 
 	// start the SQL Query receiver again
-	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort)
+	receiver, config, consumer = createTestLogsReceiverForPostgres(t, externalPort, receiverCreateSettings)
 	config.CollectionInterval = time.Second
 	config.StorageID = &storageExtension.ID
 	config.Queries = []sqlquery.Query{
@@ -276,7 +279,7 @@ func startPostgresDbContainer(t *testing.T, externalPort string) testcontainers.
 	return container
 }
 
-func createTestLogsReceiverForPostgres(t *testing.T, externalPort string) (*logsReceiver, *Config, *consumertest.LogsSink) {
+func createTestLogsReceiverForPostgres(t *testing.T, externalPort string, receiverCreateSettings receiver.CreateSettings) (*logsReceiver, *Config, *consumertest.LogsSink) {
 	factory := NewFactory()
 	config := factory.CreateDefaultConfig().(*Config)
 	config.CollectionInterval = time.Second
@@ -284,7 +287,6 @@ func createTestLogsReceiverForPostgres(t *testing.T, externalPort string) (*logs
 	config.DataSource = fmt.Sprintf("host=localhost port=%s user=otel password=otel sslmode=disable", externalPort)
 
 	consumer := &consumertest.LogsSink{}
-	receiverCreateSettings := receivertest.NewNopCreateSettings()
 	receiverCreateSettings.Logger = zap.NewExample()
 	receiver, err := factory.CreateLogsReceiver(
 		context.Background(),

--- a/receiver/sqlqueryreceiver/logs_receiver.go
+++ b/receiver/sqlqueryreceiver/logs_receiver.go
@@ -298,7 +298,7 @@ func (queryReceiver *logsQueryReceiver) collect(ctx context.Context) (plog.Logs,
 			}
 		}
 	}
-	return logs, nil
+	return logs, errors.Join(errs...)
 }
 
 func (queryReceiver *logsQueryReceiver) storeTrackingValue(ctx context.Context, row sqlquery.StringMap) error {

--- a/receiver/sqlqueryreceiver/logs_receiver.go
+++ b/receiver/sqlqueryreceiver/logs_receiver.go
@@ -6,6 +6,7 @@ package sqlqueryreceiver // import "github.com/open-telemetry/opentelemetry-coll
 import (
 	"context"
 	"database/sql"
+	"errors"
 	"fmt"
 	"time"
 
@@ -16,7 +17,6 @@ import (
 	"go.opentelemetry.io/collector/pdata/plog"
 	"go.opentelemetry.io/collector/receiver"
 	"go.opentelemetry.io/collector/receiver/receiverhelper"
-	"go.uber.org/multierr"
 	"go.uber.org/zap"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/sqlquery"
@@ -174,21 +174,21 @@ func (receiver *logsReceiver) Shutdown(ctx context.Context) error {
 		return nil
 	}
 
-	var errors error
+	var errs []error
 	receiver.settings.Logger.Debug("stopping...")
 	receiver.stopCollecting()
 	for _, queryReceiver := range receiver.queryReceivers {
-		errors = multierr.Append(errors, queryReceiver.shutdown(ctx))
+		errs = append(errs, queryReceiver.shutdown(ctx))
 	}
 
 	if receiver.storageClient != nil {
-		errors = multierr.Append(errors, receiver.storageClient.Close(ctx))
+		errs = append(errs, receiver.storageClient.Close(ctx))
 	}
 
 	receiver.isStarted = false
 	receiver.settings.Logger.Debug("stopped.")
 
-	return errors
+	return errors.Join(errs...)
 }
 
 func (receiver *logsReceiver) stopCollecting() {
@@ -286,7 +286,7 @@ func (queryReceiver *logsQueryReceiver) collect(ctx context.Context) (plog.Logs,
 		return logs, fmt.Errorf("error getting rows: %w", err)
 	}
 
-	var errs error
+	var errs []error
 	scopeLogs := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords()
 	for logsConfigIndex, logsConfig := range queryReceiver.query.Logs {
 		for _, row := range rows {
@@ -294,7 +294,7 @@ func (queryReceiver *logsQueryReceiver) collect(ctx context.Context) (plog.Logs,
 			rowToLog(row, logsConfig, logRecord)
 			logRecord.SetObservedTimestamp(observedAt)
 			if logsConfigIndex == 0 {
-				errs = multierr.Append(errs, queryReceiver.storeTrackingValue(ctx, row))
+				errs = append(errs, queryReceiver.storeTrackingValue(ctx, row))
 			}
 		}
 	}

--- a/receiver/sqlqueryreceiver/logs_receiver.go
+++ b/receiver/sqlqueryreceiver/logs_receiver.go
@@ -174,13 +174,13 @@ func (receiver *logsReceiver) Shutdown(ctx context.Context) error {
 		return nil
 	}
 
+	var errors error
 	receiver.settings.Logger.Debug("stopping...")
 	receiver.stopCollecting()
 	for _, queryReceiver := range receiver.queryReceivers {
-		queryReceiver.shutdown(ctx)
+		errors = multierr.Append(errors, queryReceiver.shutdown(ctx))
 	}
 
-	var errors error
 	if receiver.storageClient != nil {
 		errors = multierr.Append(errors, receiver.storageClient.Close(ctx))
 	}
@@ -319,5 +319,10 @@ func rowToLog(row sqlquery.StringMap, config sqlquery.LogsCfg, logRecord plog.Lo
 	logRecord.Body().SetStr(row[config.BodyColumn])
 }
 
-func (queryReceiver *logsQueryReceiver) shutdown(_ context.Context) {
+func (queryReceiver *logsQueryReceiver) shutdown(_ context.Context) error {
+	if queryReceiver.db == nil {
+		return nil
+	}
+
+	return queryReceiver.db.Close()
 }

